File: packages/dtj-typescript/test/client.test.ts
/**
 * Unit tests for client.ts
 */

import { test, describe, beforeEach, afterEach, mock } from "node:test";
import * as assert from "node:assert/strict";
import { TraceSession, NoOpTraceSession, type TraceConfig, type TraceEvent } from "../src/client.js";
import { DTJError, DTJProtocolError, DTJConnectionError, DTJAgentNotFoundError, DTJValueError, DTJSessionError } from "../src/errors.js";
import { Severity, DictKind, TypeTag, encodeValue, severityFromString } from "../src/protocol.js";

// Local type for process.emitWarning options
interface ProcessWarningOptions {
  type?: string;
  code?: string;
  detail?: string;
  constructor?: Function;
}

interface ProcessWarning extends Error {
  type?: string;
  code?: string;
  detail?: string;
}

describe("NoOpTraceSession", () => {
  test("emit is no-op but emits warning once", () => {
    const warnings: ProcessWarning[] = [];
    const originalEmitWarning = process.emitWarning;
    process.emitWarning = ((warning: string | Error, options?: ProcessWarningOptions) => {
      if (typeof warning === "string") {
        warnings.push({ message: warning, name: "Warning", ...options } as ProcessWarning);
      } else {
        warnings.push({ message: warning.message, name: warning.name, ...options } as ProcessWarning);
      }
    }) as typeof process.emitWarning;

    const session = new NoOpTraceSession();
    session.emit({ domain: "test", category: "cat", name: "event", severity: "info", fieldName: "field", value: 1 });
    session.emit({ domain: "test", category: "cat", name: "event", severity: "info", fieldName: "field", value: 2 });

    assert.equal(warnings.length, 1);
    assert.ok(warnings[0]!.message.includes("tracing disabled"));
    assert.equal(warnings[0]!.type, "DTJWarning");
    assert.equal(warnings[0]!.code, "DTJ_AGENT_UNAVAILABLE");

    process.emitWarning = originalEmitWarning;
  });

  test("close is idempotent", () => {
    const session = new NoOpTraceSession();
    session.close();
    session.close(); // Should not throw
    assert.ok(session.isClosed());
  });

  test("isClosed returns true after close", () => {
    const session = new NoOpTraceSession();
    assert.equal(session.isClosed(), false);
    session.close();
    assert.equal(session.isClosed(), true);
  });
});

describe("TraceSession", () => {
  describe("open", () => {
    test("returns NoOpTraceSession when enabled=false", async () => {
      const config: TraceConfig = {
        producerName: "test",
        producerVersion: "1.0.0",
        enabled: false,
      };
      const session = await TraceSession.open(config);
      assert.ok(session instanceof NoOpTraceSession);
    });

    test("returns NoOpTraceSession when agent not found", async () => {
      const config: TraceConfig = {
        producerName: "test",
        producerVersion: "1.0.0",
        dataDir: "/tmp/test",
        agentPath: "/nonexistent/dtj-agent",
      };
      const session = await TraceSession.open(config);
      assert.ok(session instanceof NoOpTraceSession);
    });
  });

  describe("TraceConfig", () => {
    test("accepts all required fields", () => {
      const config: TraceConfig = {
        producerName: "my-service",
        producerVersion: "1.0.0",
        dataDir: "./traces",
        agentPath: "/path/to/agent",
        socketPath: "/tmp/socket",
        sessionFileName: "custom.dtj",
        enabled: true,
      };
      assert.equal(config.producerName, "my-service");
      assert.equal(config.producerVersion, "1.0.0");
    });

    test("dataDir defaults to ./traces", () => {
      const config: TraceConfig = {
        producerName: "test",
        producerVersion: "1.0.0",
      };
      // Default is handled in open(), not in type
      assert.equal(config.dataDir, undefined);
    });

    test("enabled defaults to true", () => {
      const config: TraceConfig = {
        producerName: "test",
        producerVersion: "1.0.0",
      };
      assert.equal(config.enabled, undefined);
    });
  });

  describe("TraceEvent", () => {
    test("accepts all supported value types", () => {
      const events: TraceEvent[] = [
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: true },
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: 42n },
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: 42 },
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: 3.14 },
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: new Uint8Array([1, 2, 3]) },
        { domain: "d", category: "c", name: "n", severity: "info", fieldName: "f", value: "string", correlation: "corr" },
      ];
      assert.equal(events.length, 6);
    });
  });
});

describe("Protocol integration (mocked socket)", () => {
  // These tests verify the protocol encoding logic without a real agent
  test("encodeValue handles all supported types correctly", () => {
    // This is tested in protocol.test.ts but we verify the integration here
    
    // boolean
    let result = encodeValue(true);
    assert.equal(result.typeTag, TypeTag.BOOL);
    assert.deepEqual(result.body, new Uint8Array([1]));

    // bigint
    result = encodeValue(123n);
    assert.equal(result.typeTag, TypeTag.I64);

    // integer number
    result = encodeValue(42);
    assert.equal(result.typeTag, TypeTag.I64);

    // float number
    result = encodeValue(3.14);
    assert.equal(result.typeTag, TypeTag.F64);

    // Uint8Array
    result = encodeValue(new Uint8Array([1, 2]));
    assert.equal(result.typeTag, TypeTag.BYTES);
    assert.equal(result.body[0], 2); // length
    assert.deepEqual(result.body.slice(4), new Uint8Array([1, 2]));

    // string throws
    assert.throws(() => encodeValue("hello"), DTJValueError);
  });

  test("severityFromString maps correctly", () => {
    assert.equal(severityFromString("debug"), Severity.DEBUG);
    assert.equal(severityFromString("info"), Severity.INFO);
    assert.equal(severityFromString("warn"), Severity.WARN);
    assert.equal(severityFromString("error"), Severity.ERROR);
    assert.equal(severityFromString("fatal"), Severity.FATAL);
    assert.throws(() => severityFromString("invalid"), DTJValueError);
  });
});

describe("Error classes", () => {
  test("DTJError is base class", () => {
    const err = new DTJError("test");
    assert.ok(err instanceof Error);
    assert.equal(err.name, "DTJError");
  });

  test("DTJProtocolError includes opcode", () => {
    const err = new DTJProtocolError("protocol error", 0xFF);
    assert.equal(err.opcode, 0xFF);
    assert.equal(err.name, "DTJProtocolError");
  });

  test("DTJConnectionError", () => {
    const err = new DTJConnectionError("connection failed");
    assert.equal(err.name, "DTJConnectionError");
  });

  test("DTJAgentNotFoundError", () => {
    const err = new DTJAgentNotFoundError("agent not found");
    assert.equal(err.name, "DTJAgentNotFoundError");
  });

  test("DTJValueError", () => {
    const err = new DTJValueError("invalid value");
    assert.equal(err.name, "DTJValueError");
  });

  test("DTJSessionError", () => {
    const err = new DTJSessionError("session error");
    assert.equal(err.name, "DTJSessionError");
  });
});

File: packages/dtj-go/client_test.go
package dtj

import (
	"errors"
	"os"
	"testing"
)

func TestOpenDisabled(t *testing.T) {
	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		Enabled:         boolPtr(false),
	}
	sess := Open(cfg)
	if sess == nil {
		t.Fatalf("Open returned nil")
	}
	if !sess.closed {
		t.Fatalf("session should be closed (disabled)")
	}
	// Emit should be no-op
	if err := sess.Emit(Event{}); err != nil {
		t.Fatalf("Emit on disabled session should not error: %v", err)
	}
	// Close should be idempotent
	if err := sess.Close(); err != nil {
		t.Fatalf("Close on disabled session failed: %v", err)
	}
	if err := sess.Close(); err != nil {
		t.Fatalf("Close on disabled session second call failed: %v", err)
	}
}

func TestOpenStrictDisabled(t *testing.T) {
	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		Enabled:         boolPtr(false),
	}
	sess, err := OpenStrict(cfg)
	if err != nil {
		t.Fatalf("OpenStrict returned error: %v", err)
	}
	if sess == nil {
		t.Fatalf("OpenStrict returned nil")
	}
	if !sess.closed {
		t.Fatalf("session should be closed (disabled)")
	}
}

func TestOpenMissingAgent(t *testing.T) {
	os.Setenv("PATH", "/empty/path")
	os.Setenv("DTJ_AGENT_PATH", "")
	defer func() {
		os.Unsetenv("PATH")
		os.Unsetenv("DTJ_AGENT_PATH")
	}()

	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		DataDir:         "/tmp/test",
	}
	sess := Open(cfg)
	if sess == nil {
		t.Fatalf("Open returned nil")
	}
	if !sess.closed {
		t.Fatalf("session should be closed (no agent)")
	}
}

func TestOpenStrictMissingAgent(t *testing.T) {
	os.Setenv("PATH", "/empty/path")
	os.Setenv("DTJ_AGENT_PATH", "")
	defer func() {
		os.Unsetenv("PATH")
		os.Unsetenv("DTJ_AGENT_PATH")
	}()

	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		DataDir:         "/tmp/test",
	}
	sess, err := OpenStrict(cfg)
	if err != nil {
		t.Fatalf("OpenStrict returned error: %v", err)
	}
	if sess == nil {
		t.Fatalf("OpenStrict returned nil")
	}
	if !sess.closed {
		t.Fatalf("session should be closed (no agent)")
	}
}

func TestWarningHandler(t *testing.T) {
	var warned bool
	var warnedErr error

	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		DataDir:         "/tmp/test",
		WarningHandler: func(err error) {
			warned = true
			warnedErr = err
		},
	}

	os.Setenv("PATH", "/empty/path")
	os.Setenv("DTJ_AGENT_PATH", "")
	defer func() {
		os.Unsetenv("PATH")
		os.Unsetenv("DTJ_AGENT_PATH")
	}()

	sess := Open(cfg)
	if sess == nil {
		t.Fatalf("Open returned nil")
	}

	// Trigger warning by calling Emit
	sess.Emit(Event{
		Domain:    "test",
		Category:  "cat",
		Name:      "event",
		Severity:  SeverityInfo,
		FieldName: "field",
		Value:     1,
	})

	if !warned {
		t.Fatalf("warning handler not called")
	}
	var agentUnavail *AgentUnavailableError
	if !errors.As(warnedErr, &agentUnavail) {
		t.Fatalf("warning should be AgentUnavailableError, got: %T", warnedErr)
	}
}

func TestWarningHandlerOnce(t *testing.T) {
	warnCount := 0

	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		DataDir:         "/tmp/test",
		WarningHandler: func(err error) {
			warnCount++
		},
	}

	os.Setenv("PATH", "/empty/path")
	os.Setenv("DTJ_AGENT_PATH", "")
	defer func() {
		os.Unsetenv("PATH")
		os.Unsetenv("DTJ_AGENT_PATH")
	}()

	sess := Open(cfg)

	// Call Emit multiple times
	for i := 0; i < 5; i++ {
		sess.Emit(Event{
			Domain:    "test",
			Category:  "cat",
			Name:      "event",
			Severity:  SeverityInfo,
			FieldName: "field",
			Value:     i,
		})
	}

	if warnCount != 1 {
		t.Fatalf("warning handler called %d times, expected 1", warnCount)
	}
}

func TestEmitUnsupportedValue(t *testing.T) {
	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		Enabled:         boolPtr(false),
	}
	sess := Open(cfg)

	// On disabled session, Emit should be no-op and return nil
	err := sess.Emit(Event{
		Domain:    "test",
		Category:  "cat",
		Name:      "event",
		Severity:  SeverityInfo,
		FieldName: "field",
		Value:     struct{}{}, // unsupported type
	})

	if err != nil {
		t.Fatalf("Emit on disabled session should not error: %v", err)
	}
}

func TestCloseIdempotent(t *testing.T) {
	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
		Enabled:         boolPtr(false),
	}
	sess := Open(cfg)

	if err := sess.Close(); err != nil {
		t.Fatalf("Close failed: %v", err)
	}
	if err := sess.Close(); err != nil {
		t.Fatalf("Close second call failed: %v", err)
	}
}

func TestConfigDefaults(t *testing.T) {
	cfg := Config{
		ProducerName:    "test",
		ProducerVersion: "1.0.0",
	}

	// Enabled should default to true (nil pointer means true)
	if cfg.Enabled != nil {
		t.Fatalf("Enabled should be nil by default")
	}
}

func boolPtr(b bool) *bool {
	return &b
}
